From d972eb7aa81705c4919f608d2a9f3ba46cbf13c1 Mon Sep 17 00:00:00 2001
From: Robert Loehning <robert.loehning@qt.io>
Date: Fri, 3 Jun 2022 16:40:30 +0200
Subject: [PATCH 2/9] Check earlier to avoid sanitzer warnings

Fixes oss-fuzz issue 47689: "load of value 65, which
is not a valid value for type 'ICNSEntry::Depth'"

Pick-to: 6.4 6.3 6.2 5.15
Change-Id: Ia1b119d863e9518e308117ed1dd6a297297bc537
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit ea4684c6b17110d4ce0504f382da16462c048662)
---
 src/plugins/imageformats/icns/qicnshandler.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/plugins/imageformats/icns/qicnshandler.cpp b/src/plugins/imageformats/icns/qicnshandler.cpp
index d6fc589..1bf9074 100644
--- a/src/plugins/imageformats/icns/qicnshandler.cpp
+++ b/src/plugins/imageformats/icns/qicnshandler.cpp
@@ -462,8 +462,12 @@ static bool parseIconEntryInfo(ICNSEntry &icon)
     if (isIconCompressed(icon))
         return true;
     // Icon depth:
-    if (!depth.isEmpty())
-        icon.depth = ICNSEntry::Depth(depth.toUInt());
+    if (!depth.isEmpty()) {
+        const uint depthUInt = depth.toUInt();
+        if (depthUInt > 32)
+            return false;
+        icon.depth = ICNSEntry::Depth(depthUInt);
+    }
     // Try mono if depth not found
     if (icon.depth == ICNSEntry::DepthUnknown)
         icon.depth = ICNSEntry::DepthMono;
@@ -516,7 +520,7 @@ static bool parseIconEntryInfo(ICNSEntry &icon)
         icon.height = icon.width;
     }
     // Sanity check
-    if (icon.width == 0 || icon.width > 4096 || icon.depth > 32)
+    if (icon.width == 0 || icon.width > 4096)
         return false;
     return true;
 }
-- 
2.40.0.rc2

